include/thrust_wrapper.h

```diff
@@ -10,129 +10,292 @@
 #include <vector_thrust_allocator.h>
 #include <thrust/transform_reduce.h>
 #include <thrust/adjacent_difference.h>
+#include <type_traits>
+#include <basic_types.h>
 
 namespace thrust_wrapper
 {
-  template<typename InputIterator, typename OutputIterator>
-    inline void exclusive_scan(InputIterator first, InputIterator last, OutputIterator result)
+  template<int MemSpace, typename InputIterator, typename OutputIterator>
+    inline typename std::enable_if<MemSpace == AMGX_host, void>::type 
+    exclusive_scan(InputIterator first, InputIterator last, OutputIterator result)
     {
-      amgx::thrust_amgx_allocator<typename amgx::thrust::iterator_traits<InputIterator>::value_type> alloc;
-      amgx::thrust::exclusive_scan(amgx::thrust::cuda::par_nosync(alloc), first, last, result);
+        amgx::thrust::exclusive_scan(first, last, result);
     }
 
-  template<typename InputIterator, typename OutputIterator, typename T>
+  template<int MemSpace, typename InputIterator, typename OutputIterator>
+    inline typename std::enable_if<MemSpace == AMGX_device, void>::type 
+    exclusive_scan(InputIterator first, InputIterator last, OutputIterator result)
+    {
+        amgx::thrust_amgx_allocator<typename amgx::thrust::iterator_traits<InputIterator>::value_type> alloc;
+        amgx::thrust::exclusive_scan(amgx::thrust::cuda::par_nosync(alloc), first, last, result);
+    }
+
+  template<int MemSpace, typename InputIterator, typename OutputIterator, typename T>
     inline void exclusive_scan(InputIterator first, InputIterator last, OutputIterator result, T init)
     {
-      amgx::thrust_amgx_allocator<typename amgx::thrust::iterator_traits<InputIterator>::value_type> alloc;
-      amgx::thrust::exclusive_scan(amgx::thrust::cuda::par_nosync(alloc), first, last, result, init);
+        if(MemSpace == AMGX_host) {
+            amgx::thrust::exclusive_scan(first, last, result, init);
+        }
+        else {
+            amgx::thrust_amgx_allocator<typename amgx::thrust::iterator_traits<InputIterator>::value_type> alloc;
+            amgx::thrust::exclusive_scan(amgx::thrust::cuda::par_nosync(alloc), first, last, result, init);
+        }
```

Comment:
If i understand correctly thrust_wrapper functions have additional memoryspace parameter that for the case of `AMGX_device` will provide `amgx::thrust_amgx_allocator` device memory allocator. 

Considering that we also have pinned host memory allocator, optionally complexity of `thrust_wrapper` can be reduced by always providing either host or device memory allocator, which will be templated based on memoryspace. I.e. something like
```
template<int MemSpace, typename InputIterator, typename OutputIterator, typename T>
    inline void exclusive_scan(InputIterator first, InputIterator last, OutputIterator result, T init)
    {
      amgx::thrust_amgx_allocator<typename amgx::thrust::iterator_traits<InputIterator>::value_type, MemSpace> alloc;
      amgx::thrust::exclusive_scan(amgx::thrust::cuda::par_nosync(alloc), first, last, result, init);
    }
```
Might be just not worth the effort :)

Reply:
That seems like a good idea to me. Let me give it a try. Thanks

Reply:
I think I achieved roughly what you were thinking of in thrust_wrapper. Should be easy to extend with some compute stream too, eventually.